lib/AST/ASTContext.cpp

```diff
@@ -5032,3 +5032,203 @@ bool ASTContext::isASCIIString(StringRef s) const {
   }
   return true;
 }
+
+/// Classify usages of Self in the given type.
+///
+/// \param position The position we are currently at, in terms of variance.
+static SelfReferenceInfo
+findExistentialSelfReferences(GenericSignature existentialSig, Type type,
+                              SelfReferencePosition position) {
+  // If there are no type parameters, we're done.
+  if (!type->hasTypeParameter())
+    return SelfReferenceInfo();
+
+  // Tuples preserve variance.
+  if (auto tuple = type->getAs<TupleType>()) {
+    auto info = SelfReferenceInfo();
+    for (auto &elt : tuple->getElements()) {
+      info |= findExistentialSelfReferences(existentialSig, elt.getType(),
+                                            position);
+    }
+
+    // A covariant Self result inside a tuple will not be bona fide.
+    info.hasCovariantSelfResult = false;
+
+    return info;
+  }
+
+  // Function preserve variance in the result type, and flip variance in
+  // the parameter type.
+  if (auto funcTy = type->getAs<AnyFunctionType>()) {
+    auto inputInfo = SelfReferenceInfo();
+    for (auto param : funcTy->getParams()) {
+      // inout parameters are invariant.
+      if (param.isInOut()) {
+        inputInfo |=
+            findExistentialSelfReferences(existentialSig, param.getPlainType(),
+                                          SelfReferencePosition::Invariant);
+        continue;
+      }
+      inputInfo |= findExistentialSelfReferences(
+          existentialSig, param.getParameterType(), position.flipped());
+    }
+
+    // A covariant Self result inside a parameter will not be bona fide.
+    inputInfo.hasCovariantSelfResult = false;
+
+    auto resultInfo = findExistentialSelfReferences(
+        existentialSig, funcTy->getResult(), position);
+    if (resultInfo.selfRef == SelfReferencePosition::Covariant) {
+      resultInfo.hasCovariantSelfResult = true;
+    }
+    return inputInfo |= resultInfo;
+  }
+
+  // Metatypes preserve variance.
+  if (auto metaTy = type->getAs<MetatypeType>()) {
+    return findExistentialSelfReferences(existentialSig,
+                                         metaTy->getInstanceType(), position);
+  }
+
+  // Optionals preserve variance.
+  if (auto optType = type->getOptionalObjectType()) {
+    return findExistentialSelfReferences(existentialSig, optType, position);
+  }
+
+  // DynamicSelfType preserves variance.
+  // FIXME: This shouldn't ever appear in protocol requirement
+  // signatures.
+  if (auto selfType = type->getAs<DynamicSelfType>()) {
+    return findExistentialSelfReferences(existentialSig,
+                                         selfType->getSelfType(), position);
+  }
+
+  // Bound generic types are invariant.
+  if (auto boundGenericType = type->getAs<BoundGenericType>()) {
```

Comment:
I’m trying to figure out how it all fits together (out of sheer curiosity 😛) but I can’t seem to find the part that implements the special-case covariant `Optional.Wrapped`, `Array.Element`, and `Dictionary.Value` type parameters. As far as I can understand, this function classifies these as invariant (just like any other bound generic type). Is the special-case covariance in this (or a future) PR? Thank you, and thank you too for the helpful inline documentation!

Reply:
Special-case covariance is already implemented [on the `main` branch](https://github.com/apple/swift/blob/f3eff4a7c6c6be55a2dbe6dee0de63b2f23bad0a/lib/AST/Decl.cpp#L4889-L4924). This branch needs a rebase.